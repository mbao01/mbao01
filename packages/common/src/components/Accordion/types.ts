import * as AccordionPrimitive from "@radix-ui/react-accordion";

export type AccordionProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Root>;

export type AccordionContentProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Content
>;

export type AccordionItemProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Item>;

export type AccordionTriggerProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Trigger
>;
