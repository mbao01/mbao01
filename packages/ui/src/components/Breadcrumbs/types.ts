export type BreadcrumbsProps = {
  root?: string;
  labels?: Record<string, string>;
};
