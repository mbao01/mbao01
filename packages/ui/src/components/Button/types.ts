import type { ButtonHTMLAttributes, ReactNode } from 'react';

export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg';

export type ButtonVariant =
  | 'default'
  | 'neutral'
  | 'primary'
  | 'secondary'
  | 'accent'
  | 'ghost'
  | 'link'
  | 'info'
  | 'success'
  | 'warning'
  | 'error';

export type ButtonProps = {
  size?: ButtonSize;
  wide?: boolean;
  label: ReactNode;
  loading?: boolean;
  outline?: boolean;
  variant?: ButtonVariant;
  disabled?: boolean;
} & ButtonHTMLAttributes<HTMLButtonElement>;
