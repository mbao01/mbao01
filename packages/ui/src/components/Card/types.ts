import React from "react";

export type CardProps = {
  /**
   * applies smaller padding to the card.
   */
  compact?: boolean;
  /**
   * applies border to the card.
   */
  bordered?: boolean;
  /**
   * shows the card horizontally with image to the right or left of it.
   */
  horizontal?: boolean;
  /**
   * overlay is used to make the image full width with the content on top.
   */
  overlay?: boolean;
} & React.HTMLAttributes<HTMLDivElement>;

export type CardContentProps = React.HTMLAttributes<HTMLDivElement>;
export type CardImageProps = React.ImgHTMLAttributes<HTMLImageElement>;
export type CardHeaderProps = React.HTMLAttributes<HTMLDivElement>;
export type CardFooterProps = React.HTMLAttributes<HTMLDivElement>;
export type CardDescriptionProps = React.HTMLAttributes<HTMLParagraphElement>;
