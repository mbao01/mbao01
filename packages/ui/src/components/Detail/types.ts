import { type ReactNode } from 'react';

export type DetailProps = {
  term: string;
  children: ReactNode;
};
