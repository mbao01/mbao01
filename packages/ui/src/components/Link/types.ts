import type { To, LinkProps as RouterLinkProps } from "react-router-dom";

export type LinkVariant =
  | "default"
  | "neutral"
  | "primary"
  | "secondary"
  | "accent"
  | "link"
  | "info"
  | "success"
  | "warning"
  | "error";

export type LinkProps = Omit<RouterLinkProps, "to"> & {
  href: To;
  hover?: boolean;
  variant?: LinkVariant;
};
