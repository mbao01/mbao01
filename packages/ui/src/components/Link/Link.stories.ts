import type { Meta, StoryObj } from "@storybook/react";
import { withRouter } from "storybook-addon-react-router-v6";
import { Link } from "./Link";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Link,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withRouter],
} satisfies Meta<typeof Link>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    href: "/",
    hover: false,
    variant: "default",
    children: "Link",
  },
};

export const PrimaryLink: Story = {
  args: {
    href: "/",
    variant: "primary",
    children: "Link",
  },
};

export const ExternalLink: Story = {
  args: {
    isExternal: true,
    href: "https://google.com",
    children: "Google.com",
  },
};

export const UnderlineLink: Story = {
  args: {
    href: "/",
    underline: true,
    children: "Hover on me",
  },
};

export const HoverLink: Story = {
  args: {
    href: "/",
    hover: true,
    children: "Hover on me",
  },
};
