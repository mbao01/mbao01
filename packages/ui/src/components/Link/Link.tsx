import { Link as RouterLink } from "react-router-dom";
import { ExternalLinkIcon } from "@radix-ui/react-icons";
import { Anchor } from "@mbao01/common";
import { cn } from "@mbao01/common/utilities";
import { type LinkProps } from "./types";
import { getLinkClasses } from "./constant";

export const Link = ({
  href,
  hover,
  target,
  variant,
  children,
  className,
  underline = false,
  isExternal,
  isInternal = true,
  ...props
}: LinkProps) => {
  if (!isInternal)
    return (
      <Anchor
        href={href as string}
        target={target}
        hover={hover}
        variant={variant}
        className={className}
        underline={underline}
        isExternal={isExternal}
        {...props}
      >
        {children}
      </Anchor>
    );

  return (
    <RouterLink
      to={href}
      target={target}
      className={cn(getLinkClasses({ hover, variant, underline }), className)}
      {...props}
      {...(isExternal && { rel: "noopener noreferrer" })}
    >
      {children}
      {isExternal ? (
        <ExternalLinkIcon name="external" className="ml-[2px] inline" />
      ) : null}
    </RouterLink>
  );
};
