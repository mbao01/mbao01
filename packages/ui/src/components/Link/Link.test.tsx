import { MemoryRouter, MemoryRouterProps } from "react-router-dom";
import { render, screen } from "@testing-library/react";
import { Link } from "./Link";
import { LinkProps } from "./types";

describe("Link", () => {
  const renderLink = (props: LinkProps, memoryRouterProps?: MemoryRouterProps) => {
    return render(
      <MemoryRouter {...memoryRouterProps}>
        <Link {...props} />
      </MemoryRouter>
    );
  };

  it("renders a link with correct label", () => {
    const { asFragment } = renderLink({
      hover: false,
      children: "Go Home 🏡",
      href: { pathname: "/home", search: "hello=world", hash: "great" },
    });

    const anchorEl = screen.getByRole("link", { name: "Go Home 🏡" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "/home?hello=world#great");
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders a non internal link", () => {
    const { asFragment } = renderLink({
      isInternal: false,
      href: "https://example.com",
      children: "Google.com",
    });

    const anchorEl = screen.getByRole("link", { name: "Google.com" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "https://example.com");
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders an external link", () => {
    const { asFragment } = renderLink({
      isExternal: true,
      href: "https://example.com",
      children: "Google.com",
    });

    const anchorEl = screen.getByRole("link", { name: "Google.com" });

    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", "https://example.com");
    expect(anchorEl).toHaveAttribute("rel", "noopener noreferrer");
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    "accent",
    "default",
    "error",
    "info",
    "link",
    "neutral",
    "primary",
    "secondary",
    "success",
    "warning",
  ] as const)("has %s variant", (variant) => {
    const label = `${variant} link`;
    const { asFragment } = renderLink({
      hover: true,
      children: label,
      href: `/variant/${variant}`,
    });

    const anchorEl = screen.getByRole("link", { name: label });
    expect(anchorEl).toBeInTheDocument();
    expect(anchorEl).toHaveAttribute("href", `/variant/${variant}`);
    expect(asFragment()).toMatchSnapshot();
  });
});
