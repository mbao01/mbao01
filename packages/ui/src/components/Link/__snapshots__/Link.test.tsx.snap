// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link > has accent variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/accent"
  >
    accent link
  </a>
</DocumentFragment>
`;

exports[`Link > has default variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/default"
  >
    default link
  </a>
</DocumentFragment>
`;

exports[`Link > has error variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/error"
  >
    error link
  </a>
</DocumentFragment>
`;

exports[`Link > has info variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/info"
  >
    info link
  </a>
</DocumentFragment>
`;

exports[`Link > has link variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/link"
  >
    link link
  </a>
</DocumentFragment>
`;

exports[`Link > has neutral variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/neutral"
  >
    neutral link
  </a>
</DocumentFragment>
`;

exports[`Link > has primary variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/primary"
  >
    primary link
  </a>
</DocumentFragment>
`;

exports[`Link > has secondary variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/secondary"
  >
    secondary link
  </a>
</DocumentFragment>
`;

exports[`Link > has success variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/success"
  >
    success link
  </a>
</DocumentFragment>
`;

exports[`Link > has warning variant 1`] = `
<DocumentFragment>
  <a
    class="link transition-all link-hover"
    href="/variant/warning"
  >
    warning link
  </a>
</DocumentFragment>
`;

exports[`Link > renders a link with correct label 1`] = `
<DocumentFragment>
  <a
    class="link transition-all"
    href="/home?hello=world#great"
  >
    Go Home 🏡
  </a>
</DocumentFragment>
`;

exports[`Link > renders an external link 1`] = `
<DocumentFragment>
  <a
    class="link transition-all"
    href="https://example.com"
  >
    Google.com
    <svg
      aria-hidden="true"
      class="ml-[2px] inline h-4 w-4 align-middle"
      data-slot="icon"
      fill="currentColor"
      name="external"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M5.22 14.78a.75.75 0 0 0 1.06 0l7.22-7.22v5.69a.75.75 0 0 0 1.5 0v-7.5a.75.75 0 0 0-.75-.75h-7.5a.75.75 0 0 0 0 1.5h5.69l-7.22 7.22a.75.75 0 0 0 0 1.06Z"
        fill-rule="evenodd"
      />
    </svg>
  </a>
</DocumentFragment>
`;
